File: benchmarks/bench_data_transform.rb
# Ruby equivalent of bench_data_transform.stash
# Builds a CSV-style dataset, parses each row, groups by department,
# computes aggregates, and serializes the result back to a key=value summary.

ROWS   = 30000
PASSES = 8

def build_rows(n)
  depts = ["eng", "sales", "ops", "support", "eng", "sales"]
  names = ["Ada", "Ben", "Cyd", "Dan", "Eli", "Fay", "Gus", "Hal"]
  out = []
  i = 0
  while i < n
    d  = depts[i % 6]
    nm = names[i % 8]
    salary = 30000 + ((i * 137) % 70000)
    out << (i.to_s + "," + nm + "," + d + "," + salary.to_s)
    i += 1
  end
  out
end

puts "=== Benchmark: Data Transformation ==="

rows = build_rows(ROWS)

start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
total_summary_chars = 0
total_count = 0

PASSES.times do
  counts = {}
  sums   = {}
  maxs   = {}

  rows.each do |row|
    parts  = row.split(",")
    dept   = parts[2]
    salary = parts[3].to_i

    counts[dept] = (counts[dept] || 0) + 1
    sums[dept]   = (sums[dept] || 0) + salary

    mx = maxs[dept] || 0
    if salary > mx
      maxs[dept] = salary
    end

    total_count += 1
  end

  parts_out = []
  counts.keys.each do |k|
    c   = counts[k] || 0
    s   = sums[k] || 0
    mx  = maxs[k] || 0
    avg = s / c
    line = k + "=count:" + c.to_s +
               ",sum:"   + s.to_s +
               ",avg:"   + avg.to_s +
               ",max:"   + mx.to_s
    parts_out << line
  end
  summary = parts_out.join(";")
  total_summary_chars += summary.length
end

elapsed = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000).to_i

puts "Rows: #{ROWS}, Passes: #{PASSES}"
puts "Total rows processed: #{total_count}"
puts "Result (checksum): #{total_summary_chars}"
puts "Time: #{elapsed} ms"

File: benchmarks/bench_data_transform.pl
#!/usr/bin/perl
# Perl equivalent of bench_data_transform.stash
use strict;
use warnings;
use Time::HiRes qw(time);

my $ROWS   = 30000;
my $PASSES = 8;

sub build_rows {
    my ($n) = @_;
    my @depts = ("eng", "sales", "ops", "support", "eng", "sales");
    my @names = ("Ada", "Ben", "Cyd", "Dan", "Eli", "Fay", "Gus", "Hal");
    my @out;
    for (my $i = 0; $i < $n; $i++) {
        my $d  = $depts[$i % 6];
        my $nm = $names[$i % 8];
        my $salary = 30000 + (($i * 137) % 70000);
        push @out, $i . "," . $nm . "," . $d . "," . $salary;
    }
    return \@out;
}

print "=== Benchmark: Data Transformation ===\n";

my $rows = build_rows($ROWS);

my $start = time();
my $total_summary_chars = 0;
my $total_count = 0;

for (my $p = 0; $p < $PASSES; $p++) {
    my %counts;
    my %sums;
    my %maxs;

    for my $row (@$rows) {
        my @parts  = split(/,/, $row);
        my $dept   = $parts[2];
        my $salary = $parts[3] + 0;

        $counts{$dept} = ($counts{$dept} // 0) + 1;
        $sums{$dept}   = ($sums{$dept} // 0) + $salary;

        my $mx = $maxs{$dept} // 0;
        if ($salary > $mx) {
            $maxs{$dept} = $salary;
        }

        $total_count++;
    }

    my @parts_out;
    for my $k (keys %counts) {
        my $c   = $counts{$k} // 0;
        my $s   = $sums{$k} // 0;
        my $mx  = $maxs{$k} // 0;
        my $avg = int($s / $c);
        my $line = $k . "=count:" . $c
                      . ",sum:"   . $s
                      . ",avg:"   . $avg
                      . ",max:"   . $mx;
        push @parts_out, $line;
    }
    my $summary = join(";", @parts_out);
    $total_summary_chars += length($summary);
}

my $elapsed = int((time() - $start) * 1000);

print "Rows: $ROWS, Passes: $PASSES\n";
print "Total rows processed: $total_count\n";
print "Result (checksum): $total_summary_chars\n";
print "Time: $elapsed ms\n";
